code/weapon/weapons.cpp

```diff
@@ -6791,6 +6795,16 @@ int weapon_create( const vec3d *pos, const matrix *porient, int weapon_type, int
 		wip->animations.getAll(model_get_instance(wp->model_instance_num), animation::ModelAnimationTriggerType::OnSpawn).start(animation::ModelAnimationDirection::FWD);
 	}
 
+	if (wip->muzzle_effect.isValid()) {
+		auto particleSource = particle::ParticleManager::get()->createSource(wip->muzzle_effect);
+		particleSource.moveTo(&wp->start_pos);
+		particleSource.setOrientationMatrix(orient);
+		if(parent_objp != NULL){
+			particleSource.setVelocity(&parent_objp->phys_info.vel);
+		}
+		particleSource.finish();
+	}
+
```

Comment:
This should be alongside typical muzzle flash handling in `shipfx_flash_create`. There are a number of concerns to be dealt with which that function does, like not doing them while in cockpit view, and also imagine shotgun-type weapons, we almost always don't want duplicate effects triggered for every pellet.

Reply:
Ah, okay. I had noticed that the traditional muzzle flashes were done there and was mostly chalking it up to typical Volition weirdness; it seemed like a very strange decision to me. But that makes sense.

Reply:
How's this? Not sure if it's the best control flow.

Reply:
It's a bit uh... hard to read... you'll want to be careful about any automatic formatters, in addition to it being difficult to see the changes you actually made, it'll also make mincemeat of the blame.

Reply:
I see.

Reply:
Other than that, based on what I can tell the meaningful changes are, it looks good to me, basically how I would've done it. You might pull out
```
vec3d real_dir;
vm_vec_rotate(&real_dir, gun_dir, &objp->orient);
```
since it appears to be the same for both of them.

Reply:
All right. I wasn't sure what real_dir was for, and based on what it looked like, I was thinking it maybe rotated the direction vector according to the orientation of the weapon (which might be relevant in cases where the weapon is not being fired straight forward, e.g. with FoF).

Reply:
In both cases it's for determining the orientation of the effect (though they don't use a full orientation, they only need the forward direction) since neither the particle or muzzle effect are going to maintain any reference to the weapon itself.

Reply:
Okay, the real_dir rotation was in fact...rotating the direction upward by a few degrees for some reason? I don't really understand why, but I removed it and now it works properly.

Reply:
Oh, I didn't realize it was causing issues for you. Yeah I realize now it's turning the world direction into a local direction, since muzzle flashes are rendered local to their parent ship. But that's not true for particle effects which will mostly use world for everything.